pwnlib/encoders/i386/ascii_shellcode.py

```diff
@@ -0,0 +1,300 @@
+            vocab = b"!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
+        else:
+            required_chars = set(b'\\-%TXP')
+            allowed = set(map(ord, all_chars))
+            if not isinstance(avoid, set):
+                avoid = set(avoid)
+            if avoid.intersection(required_chars):
+                raise RuntimeError(
+                    "These characters ({}) are required because they assemble into instructions used to unpack the shellcode".format(str(required_chars, 'ascii')))
+            allowed.difference_update(avoid)
+            vocab = bytes(allowed)
+
+        if context.arch != 'i386' or context.bits != 32:
+            raise RuntimeError('Only 32-bit i386 is currently supported')
+
+        int_size = context.bits // 8
+
+        # Prepend with NOPs for the NOP sled
+        shellcode = b'\x90'*int_size + raw_bytes
+        subtractions = self._get_subtractions(shellcode, vocab)
+        allocator = self._get_allocator(len(subtractions) + self.slop, vocab)
+        nop_sled = b'P' * self.slop  # push eax
+        return allocator + subtractions + nop_sled
+
+
+    @LocalContext
+    def _get_allocator(self, size, vocab):
+        r""" Allocate enough space on the stack for the shellcode
+
+        int_size is taken from the context (context.bits / 8)
+
+        Args:
+            size (int): The allocation size
+            vocab (bytes): Allowed characters
+
+        Returns:
+            bytes: The allocator shellcode
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> encoders.i386.ascii_shellcode.encode._get_allocator(300, vocab)
+            b'TX-!!!!-!_``-t~~~P\\%!!!!%@@@@'
+        """
+        size += 0x1e # add typical allocator size
+        int_size = context.bits // 8
+        # Use eax for subtractions because sub esp, X doesn't assemble to ascii
+        result = b'TX'  # push esp; pop eax
+        # Set target to the `size` arg
+        target = struct.pack('=I', size)
+        # All we are doing here is adding (subtracting) `size` 
+        # to esp (to allocate space on the stack), so we don't care
+        # about esp's actual value. That's why the `last` parameter 
+        # for `calc_subtractions` can just be zero
+        for subtraction in self._calc_subtractions(
+            b'\x00'*int_size, target, vocab):
+            if six.PY2:
+                subtraction = str(subtraction)
+            # sub eax, subtraction
+            result += struct.pack('=c{}s'.format(int_size), b'-', subtraction)
+        result += b'P\\'  # push eax, pop esp
+        # Zero out eax for the unpacking part
+        pos, neg = self._find_negatives(vocab)
+        # and eax, pos; and eax, neg ; (0b00010101 & 0b00101010 = 0b0)
+        result += struct.pack('=cIcI', b'%', pos, b'%', neg)

```

Comment:
I think that `'<'` would be better than `'='` in `struct.pack`s, in case someone tries to pwn a x86 box from a big-endian network router or something else (which pwntools devs generally do not guarantee to work, but let's try to be portable when it costs nothing).
```suggestion
        result += struct.pack('<cIcI', b'%', pos, b'%', neg)
```

You can always use our great `packing.flat`!
```suggestion
        result += flat((b'%', pos, b'%', neg))
```